# (OpenCV - Chap5) 10월 13일
> 기본 배열 연산 (Operations on Arrays) 함수

- toc: true
- branch: master
- badges: false
- comments: true
- author: pinkocto
- categories: [python]

## 05. 기본 배열 연산 함수
OpenCV는 수학과 과학 연산을 위한 파이썬 패키지인 넘파이(numpy)와 연동해 배열을 생성할 수 있으며, 이런 배열을 처리할 수 있는 다양한 연산함수를 지원한다.

파이썬에서는 배열을 처리하기 위한 자료형으로 리스트, 튜플, 사전 등의 열거형(sequence) 객체가 있다. 리스트는 다차원의 배열을 만들고 원소를 수정할 수 있으며, 튜플은 다차원의 배열을 만들 수 있지만, 수정이 불가능한 자료형이다. OpenCV 모듈의 함수들은 넘파이 모듈의 배열(ndarray) 객체를 기반으로 입력 배열과 출력 배열을 사용한다.

이 장에서는 OpenCV에서 지원하는 여러 배열 처리 함수들을 살펴본다.

### 5.1 기본 배열 (Array) 함수

OpenCV에서는 배열을 옵션에 따라 여러 방향으로 뒤집거나 여러 번 반복하는 등 배열 자체를 처리하는 함수를 제공하고 있다. 

다음 예제는 영상파일을 읽은 후, **`cv2.flip()`, `cv2.repeat`, `cv2.transpose()`** 함수를 활용해서 상하좌우로 뒤집는 예시이다.

In [8]:
#collapse-hide
import cv2
image = cv2.imread('./ghtop_images/chap05_images/flip_test.jpg', cv2.IMREAD_COLOR)

if image is None: 
    raise Exception("영상파일 읽기 오류 발생") # 예외 처리

#### `-` 원본 이미지

In [6]:
#collapse-hide
image = cv2.imshow('image', image)

cv2.waitKey(0)
cv2.destroyAllWindows()

**<center>원본 이미지</center>**

<img src="./prac_image/image.png">

#### `-` x축 기준으로 뒤집은 이미지

In [5]:
#collapse-hide
x_axis = cv2.flip(image, 0)           # x축 기준 상하 뒤집기
cv2.imshow('x_axis', x_axis)

cv2.waitKey(0)
cv2.destroyAllWindows()

**<center>x-axis flip</center>**

<img src="./prac_image/xaxis.png">

#### `-` y축 기준으로 뒤집은 이미지

In [4]:
#collapse-hide
y_axis  = cv2.flip(image, 1)          # y축 기준 좌우 뒤집기
cv2.imshow('y_axis', y_axis)

cv2.waitKey(0)
cv2.destroyAllWindows()

**<center>y_axis flip</center>**

<img src="./prac_image/yaxis.png">

#### `-` x, y축 기준 상하좌우 뒤집기

In [3]:
#collapse-hide
xy_axis = cv2.flip(image, -1)        # 양축(x,y축) 기준 상하좌우 뒤집기
cv2.imshow('xy_axis', xy_axis)

cv2.waitKey(0)
cv2.destroyAllWindows()

**<center>xy_axis flip</center>**

<img src="./prac_image/xyaxis.png">

#### `-` 복사본 만들기

- **`cv.repeat(src, ny, nx[,dst[) -> dst`**

    - src, dst : 입력, 출력 배열

    - ny, nx : 수직, 수평방향 반복 횟수
    

- 입력 배열의 반복된 복사본으로 출력 배열을 채운다.

In [2]:
#collapse-hide
rep_image = cv2.repeat(image, 1, 2)   # 반복 복사
cv2.imshow('rep_image', rep_image)

cv2.waitKey(0)
cv2.destroyAllWindows()

**<center>repeat_image</center>**

<img src="./prac_image/repeat.png">

#### `-` 전치 이미지
- 입력 행렬의 전치 행렬을 출력으로 반환한다.

In [9]:
#collapse-hide
trans_image = cv2.transpose(image)    # 행렬 전치
cv2.imshow('trans_image', trans_image)

cv2.waitKey(0)
cv2.destroyAllWindows()

**<center>trans_image</center>**

<img src="./prac_image/trans.png">

In [11]:
#collapse-hide
image.shape

(267, 360, 3)

In [12]:
#collapse-hide
trans_image.shape 

(360, 267, 3)

$267\times 360 \times 3 \to 360\times 267\times 3$으로 전치된 것 확인!

### 5.2 채널 처리 함수

컬러 영상은 파란색(B), 녹색(G), 빨간색(R)의 각기 독립적인 2차원 정보를 합쳐 놓은 배열이라고 정의할 수 있다. 요즈음 영상처리 API에서는 컬러 영상을 표현하기 위해 채널(Channel)이라는 개념을 도입한다. 즉, 빨간색, 녹색, 파란색의 독립적인 2차원 정보는 각각 Blue채널, Green채널, Red 채널이라는 이름으로 표현된다.

다음은 단일채널 행렬을 여러 개 합치거나, 다채널을 분리하는 등 채널을 처리하는 함수에 대한 설명이다.

간단한 예제로 채널에 대한 개념을 알아보자.

- `1.` 단일채널 행렬 3개를 생성 
- `2.` 3개의 채널을 합쳐 하나의 다채널 행렬로 생성 
- `3.` 그 후 합쳐진 다채널 행렬을 다시 단일채널로 분리

In [15]:
#collapse-hide
import numpy as np
import cv2

In [16]:
#collapse-hide
## numpy.ndarray를 이용해 행렬 생성 및 초기화 방법
ch0 = np.zeros((2,4), np.uint8) + 10         # 0원소 행렬 선언 후 10 더하기
ch1 = np.ones((2,4), np.uint8) * 20          # 1원소 행렬 선언 후 20 곱하기
ch2 = np.full((2,4), 30, np.uint8)           # 행렬을 생성하며 30으로 초기화

list_bgr = [ch0, ch1, ch2]                   # 단일채널 행렬들을 모아 리스트 구성
merge_bgr = cv2.merge(list_bgr)              # 채널 합성
split_bgr = cv2.split(merge_bgr)             # 채널 분리 : 컬러영상 > 3채널 분리

In [17]:
#collapse-hide
ch0

array([[10, 10, 10, 10],
       [10, 10, 10, 10]], dtype=uint8)

In [20]:
#collapse-hide
ch1

array([[20, 20, 20, 20],
       [20, 20, 20, 20]], dtype=uint8)

In [21]:
#collapse-hide
ch2

array([[30, 30, 30, 30],
       [30, 30, 30, 30]], dtype=uint8)

In [22]:
#collapse-hide
# 단일 채널 행렬 3개 (ch0, ch1, ch2)
list_bgr 

[array([[10, 10, 10, 10],
        [10, 10, 10, 10]], dtype=uint8),
 array([[20, 20, 20, 20],
        [20, 20, 20, 20]], dtype=uint8),
 array([[30, 30, 30, 30],
        [30, 30, 30, 30]], dtype=uint8)]

In [36]:
#collapse-hide
print('merge_bgr 행렬 형태: ', merge_bgr.shape)
print(' ')
print(merge_bgr)

merge_bgr 행렬 형태:  (2, 4, 3)
 
[[[10 20 30]
  [10 20 30]
  [10 20 30]
  [10 20 30]]

 [[10 20 30]
  [10 20 30]
  [10 20 30]
  [10 20 30]]]


In [41]:
#collapse-hide
print('split_bar 행렬 형태: ', np.array(split_bgr).shape) # numpy의 shape() 함수를 적용하기 위해 ndarray객체로 변경하여 행렬 형태로 출력
print(' ')
print(split_bgr[0])
print(' ')
print(split_bgr[1])
print(' ')
print(split_bgr[2])

split_bar 행렬 형태:  (3, 2, 4)
 
[[10 10 10 10]
 [10 10 10 10]]
 
[[20 20 20 20]
 [20 20 20 20]]
 
[[30 30 30 30]
 [30 30 30 30]]


- 2열 3행 깊이가 4

In [53]:
#collapse-hide
split_bgr

(array([[10, 10, 10, 10],
        [10, 10, 10, 10]], dtype=uint8),
 array([[20, 20, 20, 20],
        [20, 20, 20, 20]], dtype=uint8),
 array([[30, 30, 30, 30],
        [30, 30, 30, 30]], dtype=uint8))

In [52]:
#collapse-hide
np.array(split_bgr)

array([[[10, 10, 10, 10],
        [10, 10, 10, 10]],

       [[20, 20, 20, 20],
        [20, 20, 20, 20]],

       [[30, 30, 30, 30],
        [30, 30, 30, 30]]], dtype=uint8)

#### 예제 실습

In [55]:
#collapse-hide
import cv2

In [56]:
image = cv2.imread('./ghtop_images/chap05_images/color.jpg', cv2.IMREAD_COLOR)  # 영상 읽기
if image is None:
    raise Exception('영상파일 읽기 오류')                  # 예외처리
    if image.ndim != 3: raise Exception("컬러 영상 아님") # 예외 처리 - 컬러 영상 확인

In [57]:
bgr = cv2.split(image)
# blue, green, red = cv2.split(image)  ## 3개 변수로 반환받기 가능!
print('bgr 자료형:',type(bgr), type(bgr[0]), type(bgr[0][0][0]))

bgr 자료형: <class 'tuple'> <class 'numpy.ndarray'> <class 'numpy.uint8'>


In [66]:
#collapse-hide
## 각 채널을 윈도우에 띄우기
cv2.imshow('image', image)
cv2.imshow('Blue chnnel', bgr[0])
cv2.imshow('Green chnnel', bgr[1])
cv2.imshow('Red chnnel', bgr[2])

cv2.waitKey(0)
cv2.destroyAllWindows()

<center>original image vs. Blue channel</center>

<img src="./prac_image/blue.png" width="700">

- Blue Channel 아래쪽 파란색 마크 부분이 Blue channel에서는 밝게 나타난다.

<center>original image vs. Green channel</center>

<img src="./prac_image/green.png" width="700">

- 원본이미지 아래쪽에 녹색을 띄는 진열대 부분이 Green Channel에서 밝게 나타난다.

<center>original image vs. red channel</center>

<img src="./prac_image/red.png" width="700">

- 원본이미지의 왼쪽 냉장 전시물의 붉은쪽 문 부분이 Red Chnnel에서 밝게 나타난다.

### 5.3.0 산술 연산 함수

행렬연산은 주로 첫 번째 배열의 i번째 원소와 두 번째 배열의 i번째 원소 간에 연산을 수행해서 결과 배열의 i번째 원소에 저장하는 방식을 취한다. 이러한 방식을 원소간 **(per-element, element-wise)** 연산이라 한다.

### 5.3.1 사칙연산

OpenCV에서 배열에 대한 사칙 연산은 두 배열의 원소간(per-element) 연산을 수행한다.


### 5.3.2 지수, 로그, 제곱근 관련 함수

OpenCV는 배열 원소의 지수와 로그 및 제곱근 관련 함수를 지원한다. 